http/transport-component.md

```diff
@@ -0,0 +1,121 @@
+# HTTP Transport Component <!-- omit in toc -->
+
+| Lifecycle Stage | Maturity      | Status | Latest Revision |
+| --------------- | ------------- | ------ | --------------- |
+| 1A              | Working Draft | Active | r0, 2023-05-31  |
+
+Authors: [@marcopolo]
+
+Interest Group: [@marcopolo], [@mxinden], [@marten-seemann]
+
+[@marcopolo]: https://github.com/marcopolo
+[@mxinden]: https://github.com/mxinden
+[@marten-seemann]: https://github.com/marten-seemann
+
+## Table of Contents <!-- omit in toc -->
+- [Context](#context)
+- [What is an HTTP transport](#what-is-an-http-transport)
+- [Multiaddr representation](#multiaddr-representation)
+- [HTTP Paths](#http-paths)
+
+
+## Context
+
+This document is only about advertising support for an HTTP transport. It
+doesn't make any assertions about how libp2p should interact with that
+transport. That will be defined in a future document.
+
+This exists to clarify the role of the `/http` component in Multiaddrs early to
+avoid confusion and conflicting interpretations.
+
+## What is an HTTP transport
+
+An HTTP transport is simply a node that can speak some standardized version of
+HTTP that is at least HTTP/1.1. Currently that means HTTP/1.1, HTTP/2, and
+HTTP/3. Intuitively if you can `curl` it with HTTP, then it speaks HTTP.
+
+Most environments will have a way to create an HTTP Client and Server, and the
+specific HTTP version used will be opaque. The client will negotiate (or [learn
+about](https://www.rfc-editor.org/rfc/rfc9114.html#section-3.1.1)) the specific HTTP version to use when communicating
+with the HTTP server.
+
+For this opaque case, we use the `/http` component at the end of the multidadr.
+If a node wants to advertise specific HTTP versions it may advertise a
+`/http-1.1`, `/http-2`, or `/h3` instead. Although these components are not
+currently [registered](https://github.com/multiformats/multiaddr), this document
+claims their use.
+
+
+## Multiaddr representation
+
+The multiaddr of a node with an HTTP transport ends with `/http` and is prefixed
+by information that would let an HTTP client know how to reach the server
+(remember that multiaddrs are [interpreted right to
+left](https://github.com/multiformats/multiaddr#interpreting-multiaddrs)). 
+
+The following are examples of multiaddrs for HTTP transport capable nodes:
+
+* `/dns/example.com/tls/http`
+* `/dns/example.com/https`
+* `/ip4/1.2.3.4/tcp/443/https`
+* `/ip4/1.2.3.4/tcp/443/tls/http`
+* `/ip6/2001:0db8:85a3:0000:0000:8a2e:0370:7334/tcp/443/tls/http`
+* `/ip4/1.2.3.4/udp/50781/tls/http` // We can infer this is an HTTP/3 address
+* `/ip4/1.2.3.4/udp/50781/quic-v1/http` // Not necessary to specify that the
```

Comment:
Let's not infer things. There might other protocols (and new QUIC versions) that also run on top of UDP.

Reply:
You're right. We shouldn't include the quic-v1 line. Just the udp and port part and the http client can figure out what HTTP transport to use (which right now only means h3).

Reply:
What do you mean? This is only for `/http` transports. To deploy a new UDP based transport you would append it to the end of a udp multiaddr (e.g. `/ip4/1.2.3.4/udp/54321/utp)

Reply:
I believe @marten-seemann's argument is that we should specify this HTTP stack is sitting on quic-v1 (rather than some other future hypothetical UDP based transport).

The general counter argument is that everything left of the `/http` component is handled by the HTTP transport. If the HTTP transport needs more information than the [authority](https://datatracker.ietf.org/doc/html/rfc3986#section-3) and whether we're using tcp vs udp, then the extra information should be supplied. Right now, I don't think there's ambiguity here. If someday there is ambiguity we can add the extra information before the `/http` component. e.g. `/ip4/1.2.3.4/udp/54321/new-transport/http`.